anndata/_core/anndata.py

```diff
@@ -90,32 +90,47 @@ def _check_2d_shape(X):
         )
 
 
+@singledispatch
 def _gen_dataframe(anno, length, index_names):
-    if isinstance(anno, pd.DataFrame):
-        anno = anno.copy()
-        if not is_string_dtype(anno.index):
-            logger.warning("Transforming to str index.")
-            anno.index = anno.index.astype(str)
-        return anno
     if anno is None or len(anno) == 0:
```

Comment:
could have registered for `type(None)` here, but since this case also applies to zero length sequences, I left it this way.